### ISSUE TITLE
RFC: assignment via integer array indexing

### ISSUE THREAD

**Issue by crusaderky** [data-apis/array-api#864]
[EDIT] this issue is about the lack of `xp.put` or equivalent `__setitem_` semantics for integar array indices. Read comments below.

[original post]
The current situation when an index is an array of ints or bools is very messy:

- numpy blindly passes every unexpected object it finds to np.asarray, which makes it accept np.ndarray, lists, tuples, but also memoryviews and anything with an `__array__` interface
- Sparse accepts lists, tuples, or numpy arrays; other sparse arrays don't work
- PyTorch won't accept numpy arrays of unsigned integers
- PyTorch won't accept PyTorch arrays of integers with dtype other than the native int
- JAX won't accept lists or tuples
- dask won't accept tuples

I think the array API standard should define a reasonable common surface.

xref https://github.com/data-apis/array-api-compat/pull/205#discussion_r1861288136

NOTE: the much more complicated case of multiple fancy indices, e.g. `a[[0, 1], [1, 2]]` is out of scope for this issue.

**Comment by rgommers** [comment]
Just a quick note that `take` is the standardized/uniform way of doing indexing with an integer array (xref gh-416 and the issue that PR links to).

You're completely right that it's very messy across libraries - it's just very hard, since we can't mandate that libraries stop doing something and break backwards compatible (or technically we can, but it'll probably just be ignored rather than implemented).

**Comment by crusaderky** [comment]
> Just a quick note that `take` is the standardized/uniform way of doing indexing with an integer array (xref [gh-416](https://github.com/data-apis/array-api/pull/416) and the issue that PR links to).

I couldn't find `put` or anything similar though? How do you update along a fancy index?

**Comment by ev-br** [comment]
cross-ref https://github.com/data-apis/array-api/pull/900/ which aims to actually standardize it.

**Comment by ev-br** [comment]
Looks like gh-900 covered this request, and we can close it as completed, I'd think.

**Comment by ev-br** [comment]
Would be great to update the issue title/test then, if we repurpose it for `__setitem__` specifically.

**Comment by amacati** [comment]
Ah right, then maybe this is a topic for array-api-extra instead. Because this code snippet should be extremely common in most frameworks:

```python
import array_api_strict as xp
import jax.numpy as jp

x = jp.array([[1, 2, 3], [4, 5, 6]])
idx = jp.array([0])
# This could be emulated in array-api-extra with xpx.at(x)[idx, ...] = xp.asarray([[7, 8, 9]])
# It is already emulating this case for regular indexing
new_x = x.at[idx, ...].set(jp.array([[7, 8, 9]]))
print(new_x)  # [[7 8 9], [4 5 6]]

x = xp.asarray([[1, 2, 3], [4, 5, 6]])
idx = xp.asarray([0])
new_x = x[idx, ...] = xp.asarray([[7, 8, 9]])  # IndexError: Fancy indexing __setitem__ is not supported.
```

**Comment by amacati** [comment]
Using the current spec, is there an array API-compatible way to implement this feature though? Because swapping the last line with 

```python
import array_api_extra as xpx
...
new_x = xpx.at(x)[idx, ...].set(xp.asarray([[7, 8, 9]]))  # IndexError: ...
```
raises the same error, since array-api-extra tries to translate this into `x[idx] = y` for array-api-strict.

**Comment by crusaderky** [comment]
> Using the current spec, is there an array API-compatible way to implement this feature though? Because swapping the last line with
> 
> import array_api_extra as xpx
> ...
> new_x = xpx.at(x)[idx, ...].set(xp.asarray([[7, 8, 9]]))  # IndexError: ...
> raises the same error, since array-api-extra tries to translate this into `x[idx] = y` for array-api-strict.

It is possible to update `xpx.at` to detect an integer array index and force an out-of-place operation in that case.

**Comment by rgommers** [comment]
> then it very well could be added to the standard, I suppose

It cannot. "real frameworks succeed" for a custom `at` function because the implementation is different per framework under the hood. And proposing a new functional `at` for the standard, something exactly zero libraries currently have, is a nonstarter.

**Comment by amacati** [comment]
> If the result is unique across all frameworks it would make sense to add a workaround for array-api-extra. If not, implementing the workaround seems misleading.

After some more digging, this is what is currently happening for frameworks:
```python
# Behavior for indexing differs between devices:
# CPU: For duplicate indices, the **last** value is used
# GPU: For duplicate indices, the **first** value is used

import cupy as cp
import jax
import jax.numpy as jnp
import numpy as np
import torch

# Numpy: CPU, last value is used
x, y = np.array([0.0, 1.0]), np.array([2.0, 3.0])
idx = np.array([1, 1])
x[idx] = y
print(x)  # [0, 3]

# Cupy: GPU, first value is used
x, y = cp.array([0.0, 1.0]), cp.array([2.0, 3.0])
x[cp.array(idx)] = y
print(x)  # [0, 2]

# Torch: CPU, last value is used
x, y = torch.tensor([0.0, 1.0]), torch.tensor([2.0, 3.0])
x[torch.tensor(idx)] = y
print(x)  # [0, 3]

# Torch: GPU, first value is used
x, y = torch.tensor([0.0, 1.0]).cuda(), torch.tensor([2.0, 3.0]).cuda()
x[torch.tensor(idx).cuda()] = y
print(x)  # [0, 2]

# JAX: CPU, last value is used
device = jax.devices("cpu")[0]
x, y = jnp.array([0.0, 1.0], device=device), jnp.array([2.0, 3.0], device=device)
x = x.at[jnp.array(idx, device=device)].set(y)
print(x)  # [0, 3]

# JAX: GPU, first value is used
x, y = jnp.array([0.0, 1.0]), jnp.array([2.0, 3.0])
x = x.at[jnp.array(idx)].set(y)
print(x)  # [0, 2]
```

The results for duplicate indices differ _depending on the device_. What's the overall sentiment on this? One could still implement special casing for array-api-strict in array-api-extra mimicking the numpy behavior, but it feels slightly misleading.

**Comment by amacati** [comment]
Alternatively, since array-api-strict is eager, we could error out on non-unique indices because of undefined behavior.

**Comment by rgommers** [comment]
> A potential long-term solution would be to be to standardise (some subset of) fancy indexing (or potentially multiple different behaviours if they are competing for standardisation??) as an extension (/extensions) to the standard. 

We only briefly touched on that, but let me just say that I think that that is unlikely. Fancy indexing is very complex, unintuitive, and hard to write down the rules for. An extension that doesn't match what libraries can implement is low value; having a _function_ that does some of what we want would be much easier to deal with.

**Comment by lucascolley** [comment]
good point, `skip_xp_backends` and `xp_capabilities` should have us covered, this is one of the reasons I advocated for this style of skip in the past
